File: Activities/ConnectingTerritories/c/connect.c
#include <stdio.h>
#include <stdlib.h>

int rows, cols, m;

int value(int r, int c) {
  int v = (r - 1) * cols + (c - 1);
  return v % m + 1;
}

int min(int a, int b) { return a < b ? a : b; }

int main() {
  scanf("%d\n%d\n%d", &rows, &cols, &m);

  int *costs = malloc(sizeof(int) * (cols + 1));
  int *costsBuffer = malloc(sizeof(int) * (cols + 1));
  int c, r;

  for (c = 1; c <= cols; c++)
    costs[c] = value(1, c);

  for (r = 2; r <= rows; r++) {
    for (c = 1; c <= cols; c++) {
      int costHere = value(r, c);
      int cost = costHere + costs[c];
      if (c > 1)
        cost = min(cost, costHere + costs[c - 1]);
      if (c < cols)
        cost = min(cost, costHere + costs[c + 1]);
      costsBuffer[c] = cost;
    }
    for (c = 1; c <= cols; c++)
      costs[c] = costsBuffer[c];
  }
  int leastCost = costs[1];
  for (c = 2; c <= cols; c++)
    leastCost = min(leastCost, costs[c]);

  printf("%d\n", leastCost);

  free(costs);
  free(costsBuffer);
  return 0;
}

File: Activities/ConnectingTerritories/python/connect.py
rows = int(input())
cols = int(input())
m = int(input())


def value(r, c):
    v = (r - 1) * cols + (c - 1)
    return v % m + 1


costs = [0] * (cols + 1)
costs_buffer = [0] * (cols + 1)
for c in range(1, cols + 1):
    costs[c] = value(1, c)
for r in range(2, rows + 1):
    for c in range(1, cols + 1):
        cost_here = value(r, c)
        cost = cost_here + costs[c]
        if c > 1:
            cost = min(cost, cost_here + costs[c - 1])
        if c < cols:
            cost = min(cost, cost_here + costs[c + 1])
        costs_buffer[c] = cost
    for c in range(1, cols + 1):
        costs[c] = costs_buffer[c]

least_cost = costs[1]
for c in range(2, cols + 1):
    if costs[c] < least_cost:
        least_cost = costs[c]
print(least_cost)
